### DISCUSSION TITLE
Running out of RAM

### DISCUSSION THREAD

**Discussion by xingjianhou** [Part-II-Computational-Physics/.github#53] [General]
When doing tasks 6 and 9, my code always works for smaller N but for N=14,15 will get something like:
`Unable to allocate 32.0 GiB for an array with shape (65536, 65536) and data type float64`
I suspect the issue is with the RAM. My system is Windows - what's the way to solve it?

**Comment by JNHallen** [comment]
32 GB is a lot of memory. Are you trying to do $N=13,\ 14$ with the CompleteHamiltonian or CompleteHamiltonian2? 

The point of using the sparse or linear operator version of the Hamiltonian is primarily to reduce the memory needed. If you are having these issues with your sparse method, that suggests you are actually creating a complete matrix – perhaps creating it first and then converting it to sparse.

**Reply by JNHallen** [reply]
Just to clarify, you should not try computing the full Hamiltonian matrix for $N > 10$.

**Reply by wchtimothy** [reply]
Is it possible to create a sparse matrix of zeros directly? For example, in order to avoid:
```
H = sparse.csr_matrix(np.zeros((2**N,2**N)))
```

**Reply by xingjianhou** [reply]
That was what I did. I printed out the zero matrix and nothing was returned - so it should be a sparse matrix. And the type() function also shows that the data type is sparse matrix. I could work with N=13,14 but not 15,16
`sparseI = sparse.csr_matrix(I)
sparseX = sparse.csr_matrix(sigmaX)
sparseZ = sparse.csr_matrix(sigmaZ)

def SparseHamiltonian(N, h=1):
    sparseH = sparse.csr_matrix(np.zeros((2**N, 2**N)))
    for i in range(1,N):
        sparseH += sparse.kron(sparse.kron(sparse.identity(2**(i-1)),sparse.csr_matrix(LocalHamiltonian(h))),sparse.identity(2**(N-i-1)))
    sparseH -= sparse.kron(sparse.kron(sparse.csr_matrix(sigmaZ),sparse.identity(2**(N-2))),sparse.csr_matrix(sigmaZ))
    sparseH -= sparse.kron(sparse.identity(2**(N-1)),sparse.csr_matrix(sigmaX)) * h /2
    sparseH -= sparse.kron(sparse.csr_matrix(sigmaX),sparse.identity(2**(N-1))) * h / 2
    return sparseH`

type(SparseHamiltonian(10)) 
#returns scipy.sparse.csr.csr_matrix`

**Reply by wchtimothy** [reply]
I think there might be an issue with creating np.zeros((2**N,2**N)) – which is big – before turning it into a sparse matrix.

**Reply by xingjianhou** [reply]
Ah I see - that makes sense. What's the way of doing it?

**Reply by wchtimothy** [reply]
I'd like to know too! At the moment I can go up to N=15 (needs more than 1 min)

**Reply by xingjianhou** [reply]
I think I found the way - I first created a sparse identity matrix and then subtracted the identity. Now it goes to 16. 
def SparseHamiltonian(N, h=1):
    sparseH = sparse.identity(2**N)
    for i in range(1,N):
        sparseH += sparse.kron(sparse.kron(sparse.identity(2**(i-1)),sparse.csr_matrix(LocalHamiltonian(h))),sparse.identity(2**(N-i-1)))
    sparseH -= sparse.kron(sparse.kron(sparse.csr_matrix(sigmaZ),sparse.identity(2**(N-2))),sparse.csr_matrix(sigmaZ))
    sparseH -= sparse.kron(sparse.identity(2**(N-1)),sparse.csr_matrix(sigmaX)) * h /2
    sparseH -= sparse.kron(sparse.csr_matrix(sigmaX),sparse.identity(2**(N-1))) * h / 2
    sparseH -= sparse.identity(2**N)
    return sparseH

**Reply by JNHallen** [reply]
Hi guys! 

You are absolutely right that the issue is that you are doing `np.zeros(2**N, 2**N)`. There are different ways of doing this, but my suggestion would be to pull one of the terms you are summing out of the loop and then creating the first matrix with kronecker products of your small matrices. Something like:

```
ii = 0
H = sparse.kron(sparse.identity(2**(ii)), sparse.kron(sparseHloc, sparse.identity(2**(N-ii-2))))
for ii in range(1, N-1):
    H += sparse.kron(sparse.identity(2**(ii)), sparse.kron(sparseHloc, sparse.identity(2**(N-ii-2))))
```

Of course you have to define sparseHloc and then add the extra terms for the periodic boundary conditions. 

Alternatively, you can create an empty sparse array by doing:
`sparse.csr_matrix((2**N, 2**))`
This has no explicitly stored elements, but scipy will assume all elements are zero if you try to add to it or operate with it.

**Reply by JNHallen** [reply]
Of course creating a sparse identity matrix and then subtracting the identity at the end is absolutely fine as a work-around!
